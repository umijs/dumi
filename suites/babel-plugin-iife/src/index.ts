import type * as BabelCore from '@babel/core';
import * as t from '@babel/types';
import {
  createDynamicImport,
  createExportObjectProperty,
  createIIFE,
  isModule,
} from './utils';

export interface DemoTransformOptions {
  /**
   * Whether to use IIFE module, the default is true
   */
  wrappedByIIFE?: boolean;
  /**
   * It can be set only when wrappedByIIFE is true,
   * which is used to specify the callee function in IIFE as an asynchronous function.
   */
  forceAsync?: boolean;
}

const ASYNC_KEY = 'async';
const EXPORTS_KEY = 'exports';

type PluginType = (
  _: typeof BabelCore,
  opts: DemoTransformOptions,
) => BabelCore.PluginObj;

const iifePlugin: PluginType = (_, opts) => {
  const { wrappedByIIFE = true, forceAsync } = opts;
  return {
    name: 'babel-plugin-iife',
    visitor: {
      Program: {
        enter(path) {
          if (!isModule(path)) return;
          if (wrappedByIIFE) {
            this.set(ASYNC_KEY, forceAsync ?? false);
          }
          this.set(EXPORTS_KEY, []);
        },
        exit(path) {
          if (!isModule(path)) return;

          const exportList = this.get(EXPORTS_KEY);
          if (exportList.length > 0) {
            path.pushContainer(
              'body',
              t.returnStatement(t.objectExpression(exportList)),
            );
          }

          path.replaceWith(
            t.program(
              wrappedByIIFE
                ? [createIIFE(path.node.body, this.get(ASYNC_KEY))]
                : path.node.body,
            ),
          );
        },
      },
      ImportDeclaration: {
        enter() {
          if (!wrappedByIIFE || forceAsync !== undefined) return;
          if (this.get(ASYNC_KEY)) return;
          this.set(ASYNC_KEY, true);
        },
        exit(path) {
          const node = createDynamicImport(path.node);
          path.replaceWith(node);
        },
      },
      ExportDeclaration: {
        exit(path) {
          const prop = createExportObjectProperty(path.node);
          if (prop) {
            this.get(EXPORTS_KEY).push(prop);
          }
          path.remove();
        },
      },
    },
  };
};

export default iifePlugin;
