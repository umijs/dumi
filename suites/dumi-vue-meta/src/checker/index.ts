export * from './Project';
export * from './TypeCheckService';
