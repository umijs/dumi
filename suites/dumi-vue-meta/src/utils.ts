import { createHash } from 'crypto';
import * as path from 'typesafe-path/posix';
import type ts from 'typescript/lib/tsserverlibrary';
import { PropertyMetaSchema, SignatureMetaSchema } from './types';

export function createNodeVisitor(
  ts: typeof import('typescript/lib/tsserverlibrary'),
) {
  return function visitNode<T extends ts.Node>(
    node: ts.Node,
    predicate: (node: ts.Node) => boolean,
  ): T | undefined {
    if (predicate(node)) {
      return node as T;
    }
    return ts.forEachChild<T>(node, (cnode: ts.Node) => {
      return visitNode<T>(cnode, predicate);
    });
  };
}

export function getNodeOfSymbol(symbol?: ts.Symbol) {
  if (symbol?.declarations?.length) {
    return symbol.declarations[0];
  }
}

export function getNodeOfType(type: ts.Type) {
  const symbol = type.aliasSymbol ?? type.symbol;
  return getNodeOfSymbol(symbol);
}

export function getJsDocTags(
  ts: typeof import('typescript/lib/tsserverlibrary'),
  typeChecker: ts.TypeChecker,
  prop: ts.Symbol | ts.Signature,
) {
  return prop.getJsDocTags(typeChecker).reduce((doc, tag) => {
    if (!doc[tag.name]) {
      doc[tag.name] = [];
    }
    doc[tag.name].push(
      tag.text !== undefined ? ts.displayPartsToString(tag.text) : '',
    );
    return doc;
  }, {} as Record<string, string[]>);
}

export function reducer(acc: any, cur: any) {
  acc[cur.name] = cur;
  return acc;
}

export function hasQuestionToken(
  ts: typeof import('typescript/lib/tsserverlibrary'),
  prop: ts.Symbol,
) {
  return prop.valueDeclaration
    ? ts.isParameter(prop.valueDeclaration) ||
      ts.isNamedTupleMember(prop.valueDeclaration)
      ? !!prop.valueDeclaration.questionToken
      : false
    : false;
}

export function getTypeOfSignature(
  typeChecker: ts.TypeChecker,
  call: ts.Signature,
) {
  const node = call.getDeclaration();
  return typeChecker.getTypeAtLocation(node);
}

export function isPromiseLike(type: ts.Type) {
  const symbol = type.getSymbol();
  return symbol?.members?.has('then' as ts.__String) || false;
}

export function getSignatureArgsMeta(
  typeChecker: ts.TypeChecker,
  subtype: ts.Type,
  argTypeTransform?: (
    node: ts.NamedTupleMember | ts.ParameterDeclaration,
  ) => PropertyMetaSchema,
) {
  const target = (subtype as ts.TypeReference).target;
  if (!target) return [];
  const labeledElementDeclarations = (target as ts.TupleType)
    .labeledElementDeclarations;
  if (!labeledElementDeclarations) return [];
  const args: SignatureMetaSchema['arguments'] = [];
  labeledElementDeclarations.forEach((node) => {
    if (!node) return;
    args.push({
      key: node.name.getText(),
      required: !node.questionToken,
      type: typeChecker.typeToString(typeChecker.getTypeAtLocation(node)),
      schema: argTypeTransform && argTypeTransform(node),
    });
  });
  return args;
}

export function signatureTypeToString(
  typeChecker: ts.TypeChecker,
  args: ts.TupleType,
  returnType: ts.Type,
) {
  const labeledElementDeclarations = args.labeledElementDeclarations || [];

  const argStringArray = labeledElementDeclarations.reduce((acc, node) => {
    if (!node) return acc;
    const typeString = typeChecker.typeToString(
      typeChecker.getTypeAtLocation(node),
    );
    const questionToken = !!node.questionToken ? '?' : '';
    acc.push(`${node.name.getText()}${questionToken}: ${typeString}`);
    return acc;
  }, [] as string[]);

  return `(${argStringArray.join(',')}) => ${typeChecker.typeToString(
    returnType,
  )}`;
}

export const BasicTypes: Record<string, string> = {
  string: 'string',
  number: 'number',
  boolean: 'boolean',
  bigint: 'bigint',
  symbol: 'symbol',
  null: 'null',
  undefined: 'undefined',
  void: 'void',
  any: 'any',
  unknown: 'unknown',
  never: 'never',
};

export function createRef(type: string, fileName: string) {
  const hash = createHash('md5');
  hash.update(`${fileName}///${type}`);
  return hash.digest('hex');
}

const windowsPathReg = /\\/g;

export function getPosixPath(anyPath: string) {
  return (anyPath as path.OsPath).replace(
    windowsPathReg,
    '/',
  ) as path.PosixPath;
}
