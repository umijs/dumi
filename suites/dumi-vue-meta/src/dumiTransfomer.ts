import { AtomComponentAsset } from 'dumi-assets-types';
import {
  FunctionPropertySchema,
  ObjectPropertySchema,
  PropertySchema,
} from 'dumi-assets-types/typings/atom/props';
import { TypeMap } from 'dumi-assets-types/typings/atom/props/types';
import type {
  ComponentLibraryMeta,
  ComponentMeta,
  EventMeta,
  MetaTransformer,
  PropertyMeta,
  PropertyMetaSchema,
  SlotMeta,
} from './types';
import { PropertyMetaKind } from './types';
import { BasicTypes } from './utils';

function getPropertySchema(schema: PropertySchema | string) {
  if (typeof schema === 'string') {
    return {
      type: schema as keyof TypeMap,
    };
  }
  return schema;
}

export const dumiTransfomer: MetaTransformer<
  Record<string, AtomComponentAsset>
> = (meta: ComponentLibraryMeta) => {
  const referencedTypes = meta.types;
  const cachedTypes: Record<string, PropertySchema | string> = {};

  function createPropertySchema(prop: PropertyMeta | EventMeta | SlotMeta) {
    const partialProp: Partial<PropertySchema> = {
      title: prop.name,
      description: prop.description,
      tags: prop.tags,
    };
    let tagDef = prop?.tags?.default;
    let def: string | undefined;
    if (tagDef?.length) {
      def = tagDef[0];
    } else if (prop.default !== undefined) {
      def = prop.default;
    }

    if (def) {
      try {
        partialProp.default = JSON.parse(def.replaceAll("'", '"'));
      } catch (error) {}
    }

    const desc = prop?.tags?.['description'];
    if (desc?.length) {
      partialProp.description = desc.join('\n');
    }
    return {
      ...partialProp,
      // eslint-disable-next-line @typescript-eslint/no-use-before-define
      ...getPropertySchema(transformSchema(prop.schema)),
    };
  }

  function transformSchema(
    schema: PropertyMetaSchema,
  ): PropertySchema | string {
    // It may not need to be checked, or it may be a basic type
    if (typeof schema === 'string') {
      const basicType = BasicTypes[schema];
      if (basicType) {
        return {
          type: basicType as any,
        };
      }
      return { type: schema as any };
    }
    switch (schema.kind) {
      case PropertyMetaKind.REF: {
        const cachedType = cachedTypes[schema.ref];
        if (cachedType) {
          return cachedType;
        }
        const type = transformSchema(referencedTypes[schema.ref]);
        cachedTypes[schema.ref] = type;
        return type;
      }
      case PropertyMetaKind.LITERAL:
        return {
          const: schema.value,
        };
      case PropertyMetaKind.BASIC:
        return {
          type: schema.type as any,
        };
      case PropertyMetaKind.ENUM:
        return {
          oneOf: (schema.schema || []).map((item) =>
            getPropertySchema(transformSchema(item)),
          ),
        };
      case PropertyMetaKind.ARRAY:
        return {
          type: 'array',
          items: schema.schema?.length
            ? getPropertySchema(transformSchema(schema.schema[0]))
            : undefined,
        };
      case PropertyMetaKind.OBJECT: {
        const required: string[] = [];
        const meta = {
          type: 'object',
          properties: Object.entries(schema.schema || {}).reduce(
            (acc, [name, prop]) => {
              if (prop.required) {
                required.push(prop.name);
              }
              acc[name] = createPropertySchema(prop);
              return acc;
            },
            {} as Record<string, PropertySchema>,
          ),
        } as ObjectPropertySchema;
        meta.required = required;
        return meta;
      }
      case PropertyMetaKind.FUNC: {
        const functionSchema = schema.schema!;
        return {
          type: 'function',
          signature: {
            isAsync: functionSchema.isAsync,
            arguments: functionSchema.arguments.map((arg) => ({
              key: arg.key,
              hasQuestionToken: !arg.required,
              type: arg.type,
            })),
            returnType: transformSchema(functionSchema.returnType),
          },
        } as FunctionPropertySchema;
      }
      case PropertyMetaKind.UNKNOWN:
        return schema.type;
    }
  }

  function transformComponent(component: ComponentMeta) {
    const { props, events, slots, exposed } = component;
    const eventsFromProps: Record<string, PropertySchema> = {};
    const required: string[] = [];
    const properties = props.reduce((acc, prop) => {
      if (prop.required) {
        required.push(prop.name);
      }
      const match = prop.name.match(/^on([A-Z].*)$/);
      if (match) {
        // Discard excluded event prop
        if (prop.schema.kind === PropertyMetaKind.UNKNOWN) {
          return acc;
        }
        const eventName = match[1].toLowerCase();
        Object.assign(prop, { name: eventName });
        eventsFromProps[eventName] = createPropertySchema(prop);
      } else {
        acc[prop.name] = createPropertySchema(prop);
      }
      return acc;
    }, {} as Record<string, PropertySchema>);

    const asset: AtomComponentAsset = {
      type: 'COMPONENT',
      propsConfig: {
        type: 'object',
        properties,
      },
      slotsConfig: {
        type: 'object',
        properties: slots.reduce((acc, slot) => {
          acc[slot.name] = createPropertySchema(slot);
          return acc;
        }, {} as Record<string, PropertySchema>),
      },
      eventsConfig: {
        type: 'object',
        properties: events.reduce((acc, event) => {
          if (acc[event.name] === undefined) {
            acc[event.name] = createPropertySchema(event);
          }
          return acc;
        }, eventsFromProps),
      },
      methodsConfig: {
        type: 'object',
        properties: exposed.reduce((acc, method) => {
          acc[method.name] = createPropertySchema(method);
          return acc;
        }, {} as Record<string, PropertySchema>),
      },
      id: component.name,
      title: component.name,
    };
    asset.propsConfig.required = required;
    return asset;
  }

  return Object.entries(meta.components).reduce((result, [name, component]) => {
    result[name] = transformComponent(component);
    return result;
  }, {} as Record<string, AtomComponentAsset>);
};
