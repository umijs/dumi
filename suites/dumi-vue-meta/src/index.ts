import {
  createParsedCommandLine,
  createParsedCommandLineByJson,
} from '@vue/language-core';
import fs from 'fs';
import process from 'process';
import * as path from 'typesafe-path/posix';
import { Project, TypeCheckService } from './checker';
import type { MetaCheckerOptions } from './types';
import { getPosixPath } from './utils';

export * from './checker';
export * from './dumiTransfomer';
export { vueTypesSchemaResolver } from './schemaResolver/custom';
export * from './types';
export { createRef } from './utils';

export type ComponentMetaChecker = typeof TypeCheckService;

export interface CheckerProjectJsonOptions {
  root: string;
  json: any;
  checkerOptions?: MetaCheckerOptions;
  ts?: typeof import('typescript/lib/tsserverlibrary');
}

/**
 * Create component metadata checker through json configuration
 */
export function createProjectByJson(options: CheckerProjectJsonOptions) {
  const {
    root,
    json,
    checkerOptions = {},
    ts = require('typescript'),
  } = options;
  const rootPath = getPosixPath(root);
  return new Project(
    () => createParsedCommandLineByJson(ts, ts.sys, root, json),
    ts,
    checkerOptions,
    rootPath,
    path.join(rootPath, 'jsconfig.json.global.vue' as path.PosixPath),
  );
}

export interface CheckerProjectOptions {
  tsconfigPath: string;
  checkerOptions?: MetaCheckerOptions;
  ts?: typeof import('typescript/lib/tsserverlibrary');
}

const defaultTsConfig = {
  compilerOptions: {
    baseUrl: './',
    strict: true,
    declaration: true,
    skipLibCheck: true,
    esModuleInterop: true,
    resolveJsonModule: true,
    jsx: 'preserve',
    jsxImportSource: 'vue',
    strictNullChecks: false,
    paths: {
      '@/*': ['src/*'],
    },
  },
  include: ['src/**/*', 'docs/**/*'],
};

/**
 * Create a meta checker for Vue project
 * @param optionsOrRootPath You can pass in the project root directory or specific configuration.
 * @example
 * ```ts
 * import { createProject } from '@dumijs/vue-meta';
 * // Manually pass in the tsconfig.json path
 * createProject({
 *   tsconfigPath: '<project-root>/tsconfig.json',
 *   checkerOptions: {},
 * });
 * ```
 * If no parameters are passed in, tsconfig.json in the current workspace will be read.
 * ```ts
 * import { createProject } from '@dumijs/vue-meta';
 * createProject();
 * ```
 */
export function createProject(
  options?: CheckerProjectOptions | string,
): Project {
  if (typeof options === 'string' || !options) {
    const rootPath = options ?? process.cwd();
    try {
      const tryPath = path.join(
        getPosixPath(rootPath),
        getPosixPath('tsconfig.json'),
      );
      fs.accessSync(tryPath, fs.constants.R_OK);
      return createProject({ tsconfigPath: tryPath });
    } catch (error) {} // ignore error
    return createProjectByJson({
      root: rootPath,
      json: defaultTsConfig,
    });
  }
  const {
    tsconfigPath,
    checkerOptions = {},
    ts = require('typescript'),
  } = options;
  const tsconfig = getPosixPath(tsconfigPath);
  return new Project(
    () => createParsedCommandLine(ts, ts.sys, tsconfigPath),
    ts,
    checkerOptions,
    path.dirname(tsconfig),
    tsconfig + '.global.vue',
  );
}
