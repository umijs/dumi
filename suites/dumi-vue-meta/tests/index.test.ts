import path from 'path';
import { afterAll, describe, expect, test } from 'vitest';
import type {
  EnumPropertyMetaSchema,
  LocalRefPropertyMetaSchema,
  MetaCheckerOptions,
  ObjectPropertyMetaSchema,
  PropertyMeta,
  TypeParamPropertyMetaSchema,
} from '../src/index';
import { createProject, vueTypesSchemaResolver } from '../src/index';
import { toRecord, tsconfigPath } from './utils';

const checkerOptions: MetaCheckerOptions = {
  forceUseTs: true,
  printer: { newLine: 1 },
  propertyResovlers: [vueTypesSchemaResolver],
  externalSymbolLinkMappings: {
    typescript: {
      Promise:
        'https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise',
    },
  },
};

const project = createProject({
  tsconfigPath,
  checkerOptions,
});

function testFeatures(kind: 'tsx' | 'sfc' | 'sfc-alias') {
  const componentPath = path.resolve(__dirname, `fixtures/${kind}/index.ts`);
  describe(`${kind}: single Vue component meta`, () => {
    describe('props', () => {
      const { component } = project.service.getComponentMeta(
        componentPath,
        'Foo',
      );
      const propMap = toRecord(component.props);

      test('option props', () => {
        const titleProp = propMap['title'];
        expect(titleProp).toMatchObject({
          required: true,
          default: "'标题'",
          type: 'string',
        });
        expect(titleProp.comment.blockTags).toContainEqual({
          tag: 'default',
          content: [{ kind: 'text', text: "''" }],
        });
      });

      test('prop referenced by assignment', () => {
        const orderProp = propMap['order'];
        expect(orderProp).toMatchObject({
          required: true,
          default: '0',
          type: 'number',
        });
      });

      test('props using PropType type assertions', () => {
        const eProp = propMap['e'] as PropertyMeta;
        expect(eProp).toMatchObject({
          required: false,
          type: 'A | 1',
        });
        expect(
          (eProp.schema as EnumPropertyMetaSchema).schema?.[1],
        ).toMatchObject({
          type: 'number',
          kind: 'literal',
          value: '1',
        });
      });

      test('external reference destructuring assignment', () => {
        const cProp = propMap['c'];
        expect(cProp).toMatchObject({
          type: '"1" | "2" | "3"',
          schema: {
            schema: [
              {
                kind: 'literal',
                type: 'string',
                value: '"1"',
              },
              {
                kind: 'literal',
                type: 'string',
                value: '"2"',
              },
              {
                kind: 'literal',
                type: 'string',
                value: '"3"',
              },
            ],
          },
        });
      });

      // ExtractPropTypes cannot infer the type of vue-types
      test.skipIf(kind === 'sfc-alias')('using vue-types', () => {
        const bProp = propMap['b'];
        expect(bProp).toMatchObject({
          type: '{ c?: string; }',
          required: true,
          default: '{}',
        });
        const dProp = propMap['d'];
        expect(dProp).toMatchObject({
          type: 'number',
          required: false,
          default: '1',
        });
      });

      test('events in props', () => {
        const onClick = propMap['onClick'];
        if (kind === 'tsx') {
          expect(onClick.schema).matchSnapshot();
          expect(onClick.description).toBe('click event');
        } else {
          // Although we do not define it in `defineProps` when writing the component,
          // we can still handle events through the onClick prop when using this component.
          // This is because vue will pass the events in defineEmits into props when processing.
          // So we can still extract onClick from the props metadata,
          // but since this is generated through `@vue/runtime-core`
          // and is within the exclude range, we can treat it as `unknown`
          // When processing the transformer, pay attention to which signature of the same attribute from props and events is better.
          expect(onClick).toMatchObject({
            schema: {
              kind: 'unknown',
            },
          });
        }
      });

      test(`async functions`, () => {
        const promiseFunc = propMap['func'];
        expect(promiseFunc.schema).toMatchObject({
          schema: { isAsync: true },
        });
      });

      test('dom type', () => {
        const dom = propMap['dom'];
        expect(dom).toMatchObject({
          schema: {
            kind: 'ref',
            externalUrl:
              'https://developer.mozilla.org/docs/Web/API/HTMLElement',
            name: 'HTMLElement',
          },
          default: 'null',
        });
      });

      test('jsdoc', () => {
        expect(propMap['func'].comment.modifierTags).toContain('alpha');
        expect(propMap['c'].comment.modifierTags).toContain('experimental');
        expect(propMap['e'].comment.modifierTags).toContain('beta');
        expect(propMap['d'].comment.blockTags).toContainEqual({
          tag: 'deprecated',
          content: [],
        });
        expect(propMap['dom'].comment.blockTags).toContainEqual({
          tag: 'since',
          content: [{ kind: 'text', text: '0.0.1' }],
        });
      });
    });

    describe('emits/events', () => {
      const { component } = project.service.getComponentMeta(
        componentPath,
        'Foo',
      );
      const eventMap = toRecord(component.events);
      test('event signature meta should be same as normal methods', () => {
        expect(eventMap['change']).matchSnapshot();
      });

      test.skipIf(kind === 'tsx')('events defined via defineEmits', () => {
        expect(eventMap['click']).matchSnapshot();
      });
    });

    describe('slots', () => {
      const { component } = project.service.getComponentMeta(
        componentPath,
        'Foo',
      );
      const slotMap = toRecord(component.slots);

      test('normal slots', () => {
        expect(slotMap['icon'].type).toBe('any');
      });

      test('scoped slots', () => {
        expect(slotMap['item']).matchSnapshot();
      });
      test('jsdoc', () => {
        expect(slotMap['item'].comment?.blockTags).toContainEqual({
          tag: 'deprecated',
          content: [],
        });
        expect(slotMap['icon'].comment?.modifierTags).toContain('experimental');
      });
    });

    describe('expose api', () => {
      const { component } = project.service.getComponentMeta(
        componentPath,
        'Foo',
      );
      const exposed = toRecord(component.exposed);
      test('ref api', () => {
        expect(Object.keys(exposed)).toStrictEqual(['focus', 'count']);
        expect(exposed['count']).toMatchObject({
          type: 'number',
        });
        expect(exposed['focus']).matchSnapshot();
      });

      test('jsdoc', () => {
        expect(exposed['count'].comment?.blockTags).toContainEqual({
          tag: 'deprecated',
          content: [{ kind: 'text', text: 'will be released in 0.0.1' }],
        });
        expect(exposed['focus'].comment?.modifierTags).toContain('alpha');
      });
    });
  });

  describe.skipIf(kind !== 'tsx')('tsx: functional component', () => {
    const { components } =
      project.service.getComponentLibraryMeta(componentPath);
    test('anonymous', () => {
      expect(components['AnonymousFComponent']).toMatchSnapshot();
    });
    test('named', () => {
      expect(components['NamedFComponent']).toMatchSnapshot();
    });
  });

  describe.skipIf(kind !== 'tsx')('tsx: composition function', () => {
    const { components, functions } =
      project.service.getComponentLibraryMeta(componentPath);
    test('composition functions', () => {
      const funcNames = Object.keys(functions);
      expect(funcNames).toStrictEqual(['useInternalValue', 'useVNode']);
      expect(functions['useInternalValue']).toMatchObject({
        kind: 'function',
        schema: {
          arguments: [
            { key: 'upstreamValue', type: '() => T' },
            { key: 'updator', type: '(upstreamValue: T, oldValue?: T) => T' },
            {
              key: 'equal',
              type: '(internalValue: T, newValue: T) => boolean',
            },
          ],
        },
      });
      expect(functions['useVNode'].type).toContain('() => VNode');
    });
    test('use @component to identify functional component', () => {
      const schema = components['InternalComponent'];
      expect(schema.props[0]).toContain({
        name: 'a',
        required: true,
        type: 'string',
      });
    });
  });

  test.skipIf(kind !== 'tsx')('generic component', () => {
    const { component, types } = project.service.getComponentMeta(
      componentPath,
      'List',
    );
    const typeParamRef = component
      .typeParams?.[0] as LocalRefPropertyMetaSchema;
    const typeParam = types[typeParamRef.ref] as TypeParamPropertyMetaSchema;
    expect(typeParam.type).toBe('Item extend BaseItem = BaseItem');
    const defaultRef = typeParam.schema?.default as LocalRefPropertyMetaSchema;
    const defaultType = types[defaultRef.ref] as ObjectPropertyMetaSchema;
    const extendRef = typeParam.schema?.type as LocalRefPropertyMetaSchema;
    const extendType = types[extendRef.ref] as ObjectPropertyMetaSchema;
    const baseItemObj = {
      id: { name: 'id', type: 'string | number' },
      text: { name: 'text' },
    };
    expect(defaultType.schema).toMatchObject(baseItemObj);
    expect(extendType.schema).toMatchObject(baseItemObj);
    // Only props of generic components can be automatically recognized
    expect(component.props).toMatchObject([
      { type: '() => BaseItem[] | Promise<BaseItem[]>', name: 'source' },
      { type: '(data: BaseItem[]) => void', name: 'onLoaded' },
    ]);
  });
}

testFeatures('tsx');
testFeatures('sfc');
testFeatures('sfc-alias');

afterAll(() => {
  project.close();
});
