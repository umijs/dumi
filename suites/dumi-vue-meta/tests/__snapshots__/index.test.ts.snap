// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sfc: single Vue component meta > emits/events > event signature meta should be same as normal methods 1`] = `
{
  "name": "change",
  "schema": {
    "kind": "function",
    "schema": {
      "arguments": [
        {
          "key": "payload",
          "required": true,
          "schema": {
            "kind": "object",
            "schema": {
              "name": {
                "description": "",
                "name": "name",
                "required": true,
                "schema": {
                  "kind": "basic",
                  "type": "string",
                },
                "tags": {},
                "type": "string",
              },
            },
            "type": "{ name: string; }",
          },
          "type": "{ name: string; }",
        },
      ],
      "isAsync": false,
      "returnType": {
        "kind": "basic",
        "type": "void",
      },
    },
    "type": "(payload: { name: string; }) => void",
  },
  "type": "(payload: { name: string; }) => void",
}
`;

exports[`sfc: single Vue component meta > emits/events > events defined via defineEmits 1`] = `
{
  "name": "click",
  "schema": {
    "kind": "function",
    "schema": {
      "arguments": [
        {
          "key": "event",
          "required": true,
          "schema": {
            "kind": "unknown",
            "type": "MouseEvent",
          },
          "type": "MouseEvent",
        },
        {
          "key": "extra",
          "required": false,
          "schema": {
            "kind": "basic",
            "type": "string",
          },
          "type": "string",
        },
      ],
      "isAsync": false,
      "returnType": {
        "kind": "basic",
        "type": "void",
      },
    },
    "type": "(event: MouseEvent,extra?: string) => void",
  },
  "type": "(event: MouseEvent,extra?: string) => void",
}
`;

exports[`sfc: single Vue component meta > expose api > ref api 1`] = `
{
  "description": "",
  "name": "focus",
  "schema": {
    "kind": "function",
    "schema": {
      "arguments": [],
      "isAsync": false,
      "returnType": {
        "kind": "basic",
        "type": "void",
      },
    },
    "type": "() => void",
  },
  "tags": {
    "description": [
      "The signature of the expose api should be obtained from here",
    ],
    "exposed": [
      "",
    ],
  },
  "type": "() => void",
}
`;

exports[`sfc: single Vue component meta > slots > scoped slots 1`] = `
{
  "description": "item",
  "name": "item",
  "schema": {
    "kind": "object",
    "schema": {
      "extra": {
        "description": "",
        "name": "extra",
        "required": false,
        "schema": {
          "kind": "basic",
          "type": "boolean",
        },
        "tags": {},
        "type": "boolean",
      },
      "list": {
        "description": "",
        "name": "list",
        "required": true,
        "schema": {
          "kind": "array",
          "schema": [
            {
              "kind": "basic",
              "type": "string",
            },
          ],
          "type": "string[]",
        },
        "tags": {},
        "type": "string[]",
      },
    },
    "type": "{ list: string[]; extra?: boolean; }",
  },
  "tags": {},
  "type": "{ list: string[]; extra?: boolean; }",
}
`;

exports[`sfc-alias: single Vue component meta > emits/events > event signature meta should be same as normal methods 1`] = `
{
  "name": "change",
  "schema": {
    "kind": "function",
    "schema": {
      "arguments": [
        {
          "key": "payload",
          "required": true,
          "schema": {
            "kind": "object",
            "schema": {
              "name": {
                "description": "",
                "name": "name",
                "required": true,
                "schema": {
                  "kind": "basic",
                  "type": "string",
                },
                "tags": {},
                "type": "string",
              },
            },
            "type": "{ name: string; }",
          },
          "type": "{ name: string; }",
        },
      ],
      "isAsync": false,
      "returnType": {
        "kind": "basic",
        "type": "void",
      },
    },
    "type": "(payload: { name: string; }) => void",
  },
  "type": "(payload: { name: string; }) => void",
}
`;

exports[`sfc-alias: single Vue component meta > emits/events > events defined via defineEmits 1`] = `
{
  "name": "click",
  "schema": {
    "kind": "function",
    "schema": {
      "arguments": [
        {
          "key": "event",
          "required": true,
          "schema": {
            "kind": "unknown",
            "type": "MouseEvent",
          },
          "type": "MouseEvent",
        },
        {
          "key": "extra",
          "required": false,
          "schema": {
            "kind": "basic",
            "type": "string",
          },
          "type": "string",
        },
      ],
      "isAsync": false,
      "returnType": {
        "kind": "basic",
        "type": "void",
      },
    },
    "type": "(event: MouseEvent,extra?: string) => void",
  },
  "type": "(event: MouseEvent,extra?: string) => void",
}
`;

exports[`sfc-alias: single Vue component meta > expose api > ref api 1`] = `
{
  "description": "",
  "name": "focus",
  "schema": {
    "kind": "function",
    "schema": {
      "arguments": [],
      "isAsync": false,
      "returnType": {
        "kind": "basic",
        "type": "void",
      },
    },
    "type": "() => void",
  },
  "tags": {
    "description": [
      "The signature of the expose api should be obtained from here",
    ],
    "exposed": [
      "",
    ],
  },
  "type": "() => void",
}
`;

exports[`sfc-alias: single Vue component meta > slots > scoped slots 1`] = `
{
  "description": "item",
  "name": "item",
  "schema": {
    "kind": "object",
    "schema": {
      "extra": {
        "description": "",
        "name": "extra",
        "required": false,
        "schema": {
          "kind": "basic",
          "type": "boolean",
        },
        "tags": {},
        "type": "boolean",
      },
      "list": {
        "description": "",
        "name": "list",
        "required": true,
        "schema": {
          "kind": "array",
          "schema": [
            {
              "kind": "basic",
              "type": "string",
            },
          ],
          "type": "string[]",
        },
        "tags": {},
        "type": "string[]",
      },
    },
    "type": "{ list: string[]; extra?: boolean; }",
  },
  "tags": {},
  "type": "{ list: string[]; extra?: boolean; }",
}
`;

exports[`tsx: single Vue component meta > emits/events > event signature meta should be same as normal methods 1`] = `
{
  "name": "change",
  "schema": {
    "kind": "function",
    "schema": {
      "arguments": [
        {
          "key": "payload",
          "required": false,
          "schema": {
            "kind": "object",
            "schema": {
              "name": {
                "description": "",
                "name": "name",
                "required": true,
                "schema": {
                  "kind": "basic",
                  "type": "string",
                },
                "tags": {},
                "type": "string",
              },
            },
            "type": "{ name: string; }",
          },
          "type": "{ name: string; }",
        },
      ],
      "isAsync": false,
      "returnType": {
        "kind": "basic",
        "type": "void",
      },
    },
    "type": "(payload?: { name: string; }) => void",
  },
  "type": "(payload?: { name: string; }) => void",
}
`;

exports[`tsx: single Vue component meta > expose api > ref api 1`] = `
{
  "description": "The signature of the expose api should be obtained from here",
  "name": "focus",
  "schema": {
    "kind": "function",
    "schema": {
      "arguments": [],
      "isAsync": false,
      "returnType": {
        "kind": "basic",
        "type": "void",
      },
    },
    "type": "() => void",
  },
  "tags": {
    "exposed": [
      "",
    ],
  },
  "type": "() => void",
}
`;

exports[`tsx: single Vue component meta > props > events in props 1`] = `
{
  "kind": "function",
  "schema": {
    "arguments": [
      {
        "key": "e",
        "required": true,
        "schema": {
          "kind": "unknown",
          "type": "MouseEvent",
        },
        "type": "MouseEvent",
      },
      {
        "key": "extra",
        "required": false,
        "schema": {
          "kind": "basic",
          "type": "string",
        },
        "type": "string",
      },
    ],
    "isAsync": false,
    "returnType": {
      "kind": "basic",
      "type": "void",
    },
  },
  "type": "(e: MouseEvent, extra?: string) => void",
}
`;

exports[`tsx: single Vue component meta > slots > scoped slots 1`] = `
{
  "description": "item",
  "name": "item",
  "schema": {
    "kind": "object",
    "schema": {
      "extra": {
        "description": "",
        "name": "extra",
        "required": false,
        "schema": {
          "kind": "basic",
          "type": "boolean",
        },
        "tags": {},
        "type": "boolean",
      },
      "list": {
        "description": "",
        "name": "list",
        "required": true,
        "schema": {
          "kind": "array",
          "schema": [
            {
              "kind": "basic",
              "type": "string",
            },
          ],
          "type": "string[]",
        },
        "tags": {},
        "type": "string[]",
      },
    },
    "type": "{ list: string[]; extra?: boolean; }",
  },
  "tags": {},
  "type": "{ list: string[]; extra?: boolean; }",
}
`;
